File: aggregator/consumer.py
import json
import logging
from kafka import KafkaConsumer
from clickhouse_driver import Client
from datetime import datetime, timezone

logging.basicConfig(level=logging.INFO)

class MetricsConsumer:
    def __init__(self, kafka_broker, kafka_topic, clickhouse_host):
        self.bootstrap_servers = kafka_broker
        self.topic = kafka_topic
        self.consumer = KafkaConsumer(self.topic, 
                                      bootstrap_servers=[kafka_broker], 
                                      value_deserializer=lambda m: json.loads(m.decode('utf-8')), 
                                      auto_offset_reset='earliest',
                                      group_id='sre-bot-aggregator')
        self.clickhouse_client = Client(host=clickhouse_host)

    def setup_table(self):
        query = """
                CREATE TABLE IF NOT EXISTS metrics (
                target String,
                body String,
                success UInt8,
                error String,
                scraped_at DateTime
                ) ENGINE = MergeTree()
                ORDER BY scraped_at
            """
        self.clickhouse_client.execute(query)

    def consume(self):
        self.consumer.subscribe([self.topic])
        try:
            for message in self.consumer:
                self.result = message.value
                scraped_at = datetime.fromisoformat(
                    self.result['ScrapedAt'].replace('Z', '+00:00')
                )

                self.clickhouse_client.execute('INSERT INTO metrics VALUES', [(
                    self.result['Target'],
                    self.result['Body'],
                    self.result['Success'],
                    self.result['Error'],
                    scraped_at)])
        except Exception as e:
            logging.error("Failed to insert message into db - %s", e)

File: aggregator/consumer.go
package main

import (
	"context"
	"encoding/json"
	"log"

	"github.com/segmentio/kafka-go"
)

type Consumer struct {
	reader *kafka.Reader
	store  *Store
	topic  string
}

func NewConsumer(broker, topic string, store *Store) *Consumer {
	kafkaReaderConfig := kafka.ReaderConfig{
		Brokers:  []string{broker},
		GroupID:  "sre-bot-aggregator",
		Topic:    topic,
		MinBytes: 1,
		MaxBytes: 10e6,
	}
	kafkaReader := kafka.NewReader(kafkaReaderConfig)
	return &Consumer{
		reader: kafkaReader,
		store:  store,
		topic:  topic,
	}
}

func (consumer *Consumer) Consume(ctx context.Context) error {
	for {

		message, err := consumer.reader.ReadMessage(ctx)
		if err != nil {
			log.Printf("Error while reading message - %v", err)
			return err
		}
		var result ScrapeResult
		if err := json.Unmarshal(message.Value, &result); err != nil {
			log.Printf("Error while unmarshalling values - %v", err)
			continue
		}
		err = consumer.store.Insert(ctx, result)
		if err != nil {
			log.Printf("Error while inserting results - %v", err)
			continue
		}
		log.Printf("Result inserted successfully for Target %s scraped at %v", result.Target, result.ScrapedAt)
	}
}

func (consumer *Consumer) Close() error {
	err := consumer.reader.Close()
	if err != nil {
		log.Printf("Error while closing reader - %v", err)
		return err
	}
	log.Print("Successfully closed reader.")
	return nil
}
